storage/backend/ceph/ceph.go

```diff
@@ -397,35 +398,35 @@ func (c *Driver) CopySnapshot(do storage.DriverOptions, snapName, newName string
 	cmd := exec.Command("rbd", "snap", "protect", intOrigName, "--snap", snapName, "--pool", do.Volume.Params["pool"])
 	er, err := runWithTimeout(cmd, do.Timeout)
 
-	if err != nil {
+	// EBUSY indicates that the snapshot is already protected.
+	if err != nil && er.ExitStatus != 0 && er.ExitStatus != int(unix.EBUSY) {
+		if er.ExitStatus == int(unix.EEXIST) {
+			err = errored.Errorf("Volume %q or snapshot name %q already exists. Snapshots cannot share the same name as the target volume.", do.Volume.Name, snapName).Combine(errors.Exists)
+		}
 		errChan <- err
 		return err
 	}
 
 	defer func() {
 		select {
 		case err := <-errChan:
-			log.Warnf("Error received while copying snapshot: %v. Attempting to cleanup.", err)
-			cmd = exec.Command("rbd", "rm", intNewName, "--pool", do.Volume.Params["pool"])
-			if er, err := runWithTimeout(cmd, do.Timeout); err != nil || er.ExitStatus != 0 {
-				log.Errorf("Error encountered removing new volume %q for volume %q, snapshot %q: %v, %v", intNewName, intOrigName, snapName, err, er.Stderr)
-				return
-			}
-			cmd := exec.Command("rbd", "snap", "unprotect", intOrigName, "--snap", snapName, "--pool", do.Volume.Params["pool"])
-			if er, err := runWithTimeout(cmd, do.Timeout); err != nil || er.ExitStatus != 0 {
-				log.Errorf("Error encountered unprotecting new volume %q for volume %q, snapshot %q: %v, %v", newName, intOrigName, snapName, err, er.Stderr)
-				return
+			if newerr, ok := err.(*errored.Error); ok && !newerr.Contains(errors.Exists) {
+				log.Warnf("Error received while copying snapshot: %v. Attempting to cleanup.", err)
+				cmd = exec.Command("rbd", "rm", intNewName, "--pool", do.Volume.Params["pool"])
+				if er, err := runWithTimeout(cmd, do.Timeout); err != nil || er.ExitStatus != 0 {
+					log.Errorf("Error encountered removing new volume %q for volume %q, snapshot %q: %v, %v", intNewName, intOrigName, snapName, err, er.Stderr)
+					return
```

Comment:
Just wondering, if the rbd clone command (below) were to fail without creating a new volume, then returning here will skip the unprotect part that follows. Is that desired?